### ISSUE TITLE
Remove dead and duplicate items feature

### ISSUE THREAD

**Issue by Oleksiy-Yakovenko** [DeaDBeeF-Player/deadbeef#504]
Original [issue 602](https://code.google.com/p/ddb/issues/detail?id=602) created by Alexey-Yakovenko on 2012-04-19T11:49:43.000Z:

submitted by: starws
Date: 2010-05-11 18:47:14

&quot;Remove dead items&quot; and &quot;Remove duplicate items&quot; features for active playlist

**Comment by wdlkmpx** [comment]
hmm this should be very easy to implement in the gtkui. 

Looking at some foobar2000 screenshots I see:
http://news.softpedia.com/images/extra/WINDOWS3/large/foobar2000/foobar2000 edit menu.png
Menu Edit
-> Remove duplicates
-> Remove dead items

Remove duplicates = coreutils 'uniq' (with absolute filenames??)
Remove dead items = verify that all files in the playlist actually exist, just one operation i suppose

When I close deadbeef and i open it again... renamed/deleted files are still in the playlist, would be really to just wipe all the offending items at the same time.. "Remove dead items"

--

Right now the right click menu of a playlist item (in deadbeef) shows "Remove" and "Remove from disk..."

Implementing this way would be really cool..
http://news.softpedia.com/images/extra/WINDOWS3/large/foobar2000/foobar2000 file operations context menu.png

**Comment by yssmcl** [comment]
Hi! Any update on this?

**Comment by DaarkWel** [comment]
Yes, it would be nice. This function existed even in olde antique Winamp :-)

**Comment by JakeSFR** [comment]
Yeah, I agree that especially "Remove dead items" would be a useful feature.

So, I was dabbling with this idea a little today and here's my attempt to implement it.
Not knowing DB's codebase and not being very proficient in C, it's just a simple "copy/paste/adjust" type of work, but, as far as I can tell after a few tests, it seems to work fine and might be at least something to start with.

Patch (against https://github.com/DeaDBeeF-Player/deadbeef/commit/fe514eeb3346ff86bd4866038cc822a8a32968db revision): [db-remove-dead-items.patch.gz](https://github.com/DeaDBeeF-Player/deadbeef/files/4008204/db-remove-dead-items.patch.gz)

Btw, one thing that bugs me is that I'd like to see this option between "Sort by" and "Configure Custom Shell Commands", but couldn't figure it out...

**Comment by Alexey-Yakovenko** [comment]
@JakeSFR Please read this: https://github.com/DeaDBeeF-Player/deadbeef/blob/master/CONTRIBUTING.md

I'm not working with patches, and I'm not finishing unfinished PRs.

**Comment by JakeSFR** [comment]
@Alexey-Yakovenko Sure, I'll make a proper PR when it's ready.
It's still WIP (for example, I just realized that it should also handle VFS), so posted it as a patch, so maybe others who are also interested in this feature will chime in.
Is that ok?

**Comment by Alexey-Yakovenko** [comment]
I thought that you posted it *instead* of PR, sorry if I misunderstood you :)

Although I think you'll have a way better chance of anybody viewing your patch if it's a PR - downloading, unpacking and opening a diff locally is not very convenient.

**Comment by JakeSFR** [comment]
No worries, now that I read my comment it wasn't 100% clear, after all. :)
Perhaps you're right about the PR, though. I'll think about it...

Anyway, here's a bit more polished revision of my previous patch:

- changed the position of the "Remove dead items" menu item to be right below the "Clear" option,

- finally figured out why I had problems with positioning the new menuitem in Edit menu, it was **./plugins/gtkui/actions.c**:
```
/* Here we have special cases for different submenus */
if (prev_title && 0 == strcmp ("File", prev_title))
    gtk_menu_shell_insert (GTK_MENU_SHELL (current), actionitem, 5);
else if (prev_title && 0 == strcmp ("Edit", prev_title))
    gtk_menu_shell_insert (GTK_MENU_SHELL (current), actionitem, 7);
else {
    gtk_container_add (GTK_CONTAINER (current), actionitem);
}
```
Changed 7 -> 8 and now it's ok,

- currently items from VFS (ZIP, RAR, etc.) are ignored, so won't be removed from the playlist even if the entire archive is missing. HTTP streams are ignored by design, btw.

So, I believe the only missing functionality is VFS.
I'll try to look into it one of these days, but of course would be nice if others could join in, too. :)

[db-remove-dead-items2.patch.gz](https://github.com/DeaDBeeF-Player/deadbeef/files/4008522/db-remove-dead-items2.patch.gz)

**Comment by JakeSFR** [comment]
@Alexey-Yakovenko: Ok, I think I nailed that VFS thing, too.
But before I'll make a PR, I have a question about **gtkui/deadbeef.glade**.
Until now I was just editing all the relevant files (callbacks, interface, etc.) manually, but was also editing that glade file, for consistency.
But I guess I should actually generate those files via glade, am I correct?

So I built the included glade-2, but when I load, even the pristine, unmodified by me, deadbeef.glade and click 'Build', the _callbacks.{c,h}_ and _support.{c,h}_ are vastly different than the original ones.
The only files it generates exactly the same way are _interface.{c,h}_.
So my question is, what exactly should I do here - should I ignore everything except interface.c and just keep the changes I made manually to _callbacks.{c,h}_?
Btw, it turned out that _interface.c_ generated by glade from my modified deadbeef.glade is exactly the same as the one I modified manually, so maybe I don't need to do anything?
Thanks.

**Comment by Alexey-Yakovenko** [comment]
Yeah, and u'll need to use the version of glade that's in the tools folder. You will need to compile it yourself, other versions won't work.

**Comment by JakeSFR** [comment]
Yeah, I was already using the included glade.
Anyway, the PR is there (https://github.com/DeaDBeeF-Player/deadbeef/pull/2299) for you to review it.
I have absolutely no idea if I've done it the right way, though... :)

**Comment by Alexey-Yakovenko** [comment]
I'm on a trip with no computer for the next 10 days, will look at it when I'm back :-) happy new year!

**Comment by JakeSFR** [comment]
Sure, there's no rush. Happy New Year!

**Comment by kpcee** [comment]
I have implemented a similar feature but have not yet made a Pull Request. With this you can synchronize a playlist with a folder and its subdirectories, vanished files will be removed from the playlist, new songs will be added but duplicates are not considered.
![Unbenannt](https://user-images.githubusercontent.com/39984895/71641640-d567e700-2c9f-11ea-89b3-4d67f0acc9d4.png)

**Comment by ldexterldesign** [comment]
Discovered this project today after itunes/music > quod libet > deadbeef and looks good!

As a new user duplicate handling was one of the first use cases I tried, be great to avoid this:

<img width="1440" alt="Screenshot 2020-01-04 at 14 15 40" src="https://user-images.githubusercontent.com/1715809/71766153-26c0e200-2efd-11ea-9578-e31425e57f72.png">

Keep up great work!

Yours hopefully

**Comment by kpcee** [comment]
Instead of adding a function to delete duplicates, I would prefer a way to sort out such files as soon as the user adds them. Without an option in the preferences dialog, it would be less than 10 lines of code to realise this. Just my two cents.

**Comment by kpcee** [comment]
Another nice way to add this functionality is via plugin [Playlist Manager](https://github.com/kpcee/deadbeef-playlist-manager).
